infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -113,6 +108,7 @@ public LogReplicationSinkManager(String localCorfuEndpoint, LogReplicationConfig
          */
         this.rxState = RxState.LOG_ENTRY_SYNC;
         this.config = config;
+        this.topologyConfigId = topologyConfigId;
```

Comment:
I have a test to verify this fix. It will come as part of another PR as it has refactored the CorfuLogReplicationE2ETest to make the methods as utils for other tests.

Reply:
The clusterManager should query the config and update the config ID properly. Does the clusterManager not start to cause the issue?

Reply:
The clusterManager does update the persisted topologyConfigId. But within the LogReplicationSinkManager we use the cached value (see for instance the receive method). This cached value is updated on processTopologyConfigChange but not in the initial path (when there is no change but we're starting)... So topologyConfigId on the first run is always 0 instead of the one received by the ClusterManager... We're ok on the update path but fail to set it on the start path...

Reply:
Then it is better that sinkManager doesn't cache the configureID. If the sinkManager needs the configID, it just queries the metadata table to get the most recent one.

Reply:
For every received message we verify it is intended for us on that 'epoch'. That would mean accessing the database for every received message.

Reply:
The data is at runtime, it is a in-memory database, the accessing should be fast. Just one scenario: a node missed the notification of the clusterConfig change due to network partition, this node join the cluster again with the old config and becomes the leader later, will its cached configID be out of date?

Reply:
One question, why is it an in-memory database? Isn't the metadata table manager doing an access to Corfu itself? which is not running in-memory and which could be doing an RPC call by reading from the tail.

Maybe this scenario that you're highlighting is an indicative that on leadership (lock acquisition) we should re-fetch the topology, cause what if all 3 nodes missed the notification? even if we read from the datastore it is going to rely on an old config. Maybe that's the second change that we should do.

Reply:
If performance is the concern, we can remove the epoch and epoch check at SinkManager, the snapshotWriter/logEntryWriter will do the check anyway. The checking at sinkManager just brings more complexity. It is nice to pass the epoch down at the startup to the metadata manager.

Reply:
But that is just moving the problem one layer up, right? cause still we process every message, so for every message we would be accessing the database. 

I would say that maybe we should keep the topologyConfigId cached and it is refreshed on two events:
- On lock acquisition (we need to add this, to avoid the scenario of all nodes missing notification)
- On topology change notification

Reply:
The writers always do the check no matter what.

Reply:
But it is a possible case. The writers can continue to do the check always, that is not in question. I'm just concerned of having to access the database on every apply (every message) to check the topologyConfigId. TopologyConfigId changes anyways always come from the discovery service, either because:
(1) the lock was acquired and we check what is the recent topology  (to be added), or
(2) because a changeOfTopologyNotification came in

In this sense, we can always push this value down. Anyways reading from the DB is because we persisted it from the discovery service on (1) or (2) and instead of pushing the new topologyConfigId down, down there we read it from the database. There is where I don't think we need to be doing RPC calls for every received message if we can have it locally by being pushed by the Discovery Service.

Reply:
I think adding a checking at SinkManager is ok. Just when the message has a higher topology epoch, the current node should send an alarm to DiscoveryService to get the most recent topology information etc. We can address that in another PR.